lua/gatekeeper/staticlib.lua

```diff
@@ -74,6 +82,34 @@ function alloc_an_lcore (numa_table)
 	return lcore_t[1]
 end
 
+function merge_arrays (a, b)
+	for i, v in ipairs(b) do
+		table.insert(a, v)
+	end
+	return a
+end
+
+function alloc_lcores_evenly_from_all_numa_nodes (numa_table, n)
+	local num_numa_nodes = get_num_numa_nodes (numa_table)
+	local q = n / num_numa_nodes
+	local r = n % num_numa_nodes
+	local i = 0
+	local res = {}
+	for numa, lcores in all_ipairs(numa_table) do
+		local lcores_needed = q + ((i < r) and 1 or 0)
+		if #lcores >= lcores_needed then
+			local a1, a2 = split_array(lcores, lcores_needed)
+			numa_table[numa] = a2
+			numa_table["__net_conf"].numa_used[numa] = true
+			res = merge_arrays(res, a1)
+		else
+			error("There is not enough lcores");
+		end
+		i = i + 1
+	end
+	return res
```

Comment:
Instead of returning an array of lcores, `alloc_lcores_evenly_from_all_numa_nodes()` should return a table analogous to `numa_table`, that is, a hash table whose key is the NUMA node and the value an array of lcores. This change makes it easier to remove the lcores for the SOL blocks: call `alloc_lcores_evenly_from_all_numa_nodes()` with the return of the first call and the number of NUMA nodes in the table of the first return.

Of course, we'll need a function to convert the return of `alloc_lcores_evenly_from_all_numa_nodes()` into an array.